geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/ingest/IngestCommand.scala

```diff
@@ -238,47 +238,89 @@ object IngestCommand extends LazyLogging {
-      val file = inputs.iterator.flatMap(PathUtils.interpretPath).headOption.getOrElse {
-        throw new ParameterException("Parameter <files> did not evaluate to anything that could be read")
+      var open: () => InputStream = null
+      var file: FileHandle = null
+
+      // try to infer the schema from stdin
+      if (inputs == Inputs.StdInInputs) {
+        Command.user.info("No converter defined - will attempt to detect schema from stdin")
+
+        // Cache the bytes from the input stream, because we'll need them again for the ingest
+        val inputBytes = System.in.readAllBytes()
+
+        // Function that returns an input stream containing the data from stdin
+        def openStdin(): InputStream = {
+          try {
+            // This prevents the schema and type inference code from consuming all the bytes from stdin and leaving
+            // no bytes for the actual ingest to read
+            System.setIn(new ByteArrayInputStream(inputBytes))
+
+            // Return stdin
+            System.in
+          } catch {
+            case _: Exception => throw new ParameterException("Unable to read from stdin")
+          }
+        }
+
+        open = openStdin
       }
-      val opened = ListBuffer.empty[CloseableIterator[InputStream]]
-      def open(): InputStream = {
-        val streams = file.open.map(_._2)
-        opened += streams
-        if (streams.hasNext) { streams.next } else {
+      // try to infer the schema from the input files themselves
+      else {
+        Command.user.info("No converter defined - will attempt to detect schema from input files")
+        file = inputs.iterator.flatMap(PathUtils.interpretPath).headOption.getOrElse {
           throw new ParameterException("Parameter <files> did not evaluate to anything that could be read")
         }
+        val opened = ListBuffer.empty[CloseableIterator[InputStream]]
+
+        def openFile(): InputStream = {
+          val streams = file.open.map(_._2)
+          opened += streams
+          if (streams.hasNext) {
+            streams.next
+          } else {
+            throw new ParameterException("Parameter <files> did not evaluate to anything that could be read")
+          }
+        }
+
+        open = openFile
       }
+
       val (inferredSft, inferredConverter) = try {
         val opt = format match {
-          case None      => SimpleFeatureConverter.infer(open, Option(sft), Option(file.path))
-          case Some(fmt) => TypeAwareInference.infer(fmt, open, Option(sft), Option(file.path))
+          case None => SimpleFeatureConverter.infer(open, Option(sft))
+          case Some(fmt) => TypeAwareInference.infer(fmt, open, Option(sft), None)
         }
         opt.getOrElse {
           throw new ParameterException("Could not determine converter from inputs - please specify a converter")
         }
       } finally {
-        CloseWithLogging(opened)
+        CloseWithLogging(open())
```

Comment:
I think this line needs to be reverted, and the `opened` array needs to be moved up out of the if block so it stays in scope. The way this is currently working is that you're opening up new InputStreams and then closing them, while the input streams opened by type inference do not get closed. Normally the inferencer code that opens the stream would be responsible for closing it, but we need this extra layer of cleanup b/c some files can evaluate to multiple inputs (i.e. a zip file), and the inferencer isn't aware of the outer file that needs to be closed.